File: app/src/main/java/com/enginebai/poc/MyApplication.java
package com.enginebai.poc;

import android.app.Application;

import com.enginebai.core.CoreModule;
import com.enginebai.poc.data.user.UserDataHelper;
import com.enginebai.poc.delegate.CoreApp;
import com.enginebai.poc.di.AppComponent;
import com.enginebai.poc.di.AppInjector;
import com.enginebai.poc.di.DomainComponent;
import com.enginebai.poc.di.HasSingletonComponent;
import com.enginebai.poc.di.MySingletonComponent;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import dagger.hilt.EntryPoints;
import dagger.hilt.android.HiltAndroidApp;

/**
 *  1. UI Flow:
 *  MainActivity + MainViewModel
 *  |-- SingleFragmentsActivity
 *  |---- SingletonFragment 1
 *  |---- SingletonFragment ..
 *  |---- SingletonFragment n
 *  |------ SingletonDetailActivity
 *  |
 *  |-- DomainActivity
 *  |-- DomainFragmentsActivity
 *  |---- DomainFragment + DomainFragmentViewModel
 *
 *  2. Dependency Graph:
 * MyApplication -> Application -> MainViewModel (AndroidViewModel)
 *
 * MyApplication -> Context -> UserDataHelper
 * Calendar -> UserDataHelper
 * DateFormat -> UserDataHelper
 * UUID -> UserDataHelper
 * Random -> UserDataHelper
 *
 * RGB -> AppColor -> AppComponent.appColor() -> MainActivity.backgroundColor
 * RGB -> SingletonColor -> SingletonComponent interface
 *      -> SingleFragment / SingletonDetailActivity.backgroundColor
 */
@HiltAndroidApp
public class MyApplication extends Application implements HasSingletonComponent {

    // Accessing the interfaces
//    private AppComponent appComponent;
//    private DomainComponent domainComponent;

    @Inject
    UserDataHelper userDataHelper;

    // For dagger.android
//    @Inject
//    DispatchingAndroidInjector<Object> dispatchingAndroidInjector;
    @Inject
    AppInjector appInjector;

//    @Override
//    public AndroidInjector<Object> androidInjector() {
//        return dispatchingAndroidInjector;
//    }

    @Override
    public void onCreate() {
        super.onCreate();

        initDelegate();

        // call function from injected filed
        userDataHelper.generateNewUser();

        instantiateDomainComponent();
        appInjector.init(this);
    }

    // onConfigAvailable() method
    public void instantiateDomainComponent() {
//        domainComponent = appComponent.plus(new DomainModule());
//        domainComponent.inject(this);
    }

    public AppComponent appComponent() {
        return EntryPoints.get(this, AppComponent.class);
    }

    public DomainComponent domainComponent() {
        return EntryPoints.get(this, DomainComponent.class);
    }

    public void initDelegate() {
        CoreModule.INSTANCE.init(new CoreApp(this));
    }

    @NonNull
    @Override
    public MySingletonComponent getSingletonComponent() {
        return appComponent();
    }
}

File: app/src/main/java/com/enginebai/poc/MyApplication.kt
package com.enginebai.poc

import android.app.Application
import com.enginebai.core.CoreModule.init
import com.enginebai.core.Initializer
import com.enginebai.poc.data.user.UserDataHelper
import com.enginebai.poc.delegate.CoreApp
import com.enginebai.poc.di.*
import dagger.android.AndroidInjector
import dagger.android.DispatchingAndroidInjector
import dagger.android.HasAndroidInjector
import org.koin.android.ext.android.inject
import javax.inject.Inject

/**
 * 1. UI Flow:
 * MainActivity + MainViewModel
 * |-- SingleFragmentsActivity
 * |---- SingletonFragment 1
 * |---- SingletonFragment ..
 * |---- SingletonFragment n
 * |------ SingletonDetailActivity
 * |
 * |-- DomainActivity
 * |-- DomainFragmentsActivity
 * |---- DomainFragment + DomainFragmentViewModel
 *
 * 2. Dependency Graph:
 * MyApplication -> Application -> MainViewModel (AndroidViewModel)
 *
 * MyApplication -> Context -> UserDataHelper
 * Calendar -> UserDataHelper
 * DateFormat -> UserDataHelper
 * UUID -> UserDataHelper
 * Random -> UserDataHelper
 *
 * RGB -> AppColor -> AppComponent.appColor() -> MainActivity.backgroundColor
 * RGB -> SingletonColor -> SingletonComponent interface
 * -> SingleFragment / SingletonDetailActivity.backgroundColor
 */
class MyApplication : Application(), HasAndroidInjector {
    // Accessing the interfaces
    private lateinit var appComponent: AppComponent
    private lateinit var domainComponent: DomainComponent

    private val userDataHelper: UserDataHelper by inject()

    @Inject
    lateinit var appInjector: AppInjector

    // For dagger.android
    @Inject
    lateinit var dispatchingAndroidInjector: DispatchingAndroidInjector<Any>

    override fun androidInjector(): AndroidInjector<Any> {
        return dispatchingAndroidInjector
    }

    override fun onCreate() {
        super.onCreate()
        appComponent = DaggerAppComponent
            .builder()
            .application(this)
            .context(this)
            .build()
        appComponent.inject(this)
        initDelegate()
        appComponent.koinFacade()

        // call function from injected filed
        userDataHelper.generateNewUser()
        instantiateDomainComponent()
        appInjector.init(this)
//        initDynamicFeatureModule()
    }

    // onConfigAvailable() method
    private fun instantiateDomainComponent() {
        domainComponent = appComponent.plus(DomainModule())
        domainComponent.inject(this)
        domainComponent.koinFacade().loadModules()
    }

    fun changeDomain() {
        domainComponent.koinFacade().unloadModules()
        instantiateDomainComponent()
    }

    fun appComponent(): AppComponent {
        return appComponent
    }

    fun domainComponent(): DomainComponent {
        return domainComponent
    }

    private fun initDelegate() {
        init(CoreApp(this))
    }

    private fun initDynamicFeatureModule() {
        val implementationName = "com.enginebai.dynamic.DynamicFeatureInitializer"
        val initializer: Initializer = Class.forName(implementationName).newInstance() as Initializer
        initializer.initialize(this)
    }
}